shiny/playwright/controller/_base.py

```diff
@@ -0,0 +1,443 @@
+      or `expect_label_to_have_attribute()` methods. Just use `expect_label()`
+  * There are no properties, only methods; This allows for timeout values to be passed through and for complex methods.
+    * Locators will stay as properties
+  * Don't sub-class. For now, use `_` separatation and use `loc` or `value` as a prefix
+* Approach
+  * Use locators / playwright_expect as much as possible
+    * It should not be necessary to use `assert` directly.
+    * MUST wait for `Locator`s to do their job
+  * DO NOT provide `value` methods
+  * Add _set_ methods (or set like methods) only if a user would perform them
+"""
+
+"""
+# Mixins
+* Use mixins to add consistent functionality to different classes
+* These classes should **never** be instantiated directly
+* Use `typing.Protocol` to define the interface of what is required on `self`
+* Add methods to the mixin if they are consistently used across multiple classes
+  * If a method is only used in one class, it should be defined in that class
+  * If a method is used inconsistently, make/use a helper method
+"""
+
+InitLocator = typing.Union[Locator, str]
+
+R = typing.TypeVar("R")
+M1 = typing.TypeVar("M1")
+M2 = typing.TypeVar("M2")
+
+
+def is_missing(x: object) -> TypeGuard[MISSING_TYPE]:
+    return isinstance(x, MISSING_TYPE)
+
+
+# TypeGuard does not work for `not isinstance(x, MISSING_TYPE)`
+# See discussion for `StrictTypeGuard`: https://github.com/python/typing/discussions/1013
+# Until then, we need `not_is_missing(x=)` to narrow within an `if` statement
+def not_is_missing(x: R | MISSING_TYPE) -> TypeGuard[R]:
+    return not isinstance(x, MISSING_TYPE)
+
+
+def all_missing(*args: object) -> TypeGuard[MISSING_TYPE]:
+    for arg in args:
+        if not_is_missing(arg):
+            return False
+    return True
+
+
+def maybe_missing(x: M1 | MISSING_TYPE, default: M2) -> M1 | M2:
+    if isinstance(x, MISSING_TYPE):
+        return default
+    return x
+
+
+def set_text(
+    loc: Locator,
+    text: str,
+    *,
+    delay: OptionalFloat = None,
+    timeout: Timeout = None,
+) -> None:
+    """
+    Sets the text of a DOM element.
+
+    Parameters
+    ----------
+    loc
+        Playwright `Locator` of the element.
+    text
+        The text to set.
+    delay
+        The delay between key presses in milliseconds. Defaults to `None`.
+    timeout
+        The maximum time to wait for the text to be set. Defaults to `None`.
+    """
+    # TODO-future; Composable set() method
+    loc.fill("", timeout=timeout)  # Reset the value
+    loc.type(text, delay=delay, timeout=timeout)  # Type the value
+
+
+def _expect_multiple(loc: Locator, multiple: bool, timeout: Timeout = None) -> None:
+    value = "True" if multiple else None
+    _expect_style_to_have_value(loc, "multiple", value, timeout=timeout)
+
+
+######################################################
+# # Inputs
+######################################################
+
+
+class _UiBaseP(Protocol):
```

Comment:
Please rename all classes below by removing their leading `_`.

```python
class UiBaseP(Protocol):
```

The class names (or any other variables) should not start with `_` as they are used outside this module.